hamamatsu.py

```diff
@@ -134,418 +177,402 @@ def set_by_dict(attr, node_text, values):
+        not_enough_buffers = bf_dif > self.num_img_buffers
+
+        # TODO : use logging.debug
+        print(f"Last Frame : {self.last_frame_acquired}\n"
+              f"New Frame : {last_buf_num}\n"
+              f"Difference : {bf_dif}")
+        if not session_acquiring:
+            er_msg = "In session.status() NOT acquiring."
+            raise SomeError(er_msg)  # TODO : Replace placeholder
+        if last_buf_num != self.last_frame_acquired and last_buf_num != -1 and not_enough_buffers:
+            er_msg = "The number of images taken exceeds the number of buffers alloted." + \
+                "Images have been lost.  Increase the number of buffers."
+            raise SomeError(er_msg)  # TODO : Replace placeholder
+            # TODO : Use logger
+
+        frame_ind = self.last_frame_acquired
+        for i in range(bf_dif):
+            frame_ind += 1
+            # Why is this in the labview code? Should be a flag for you verbose logging is maybe?
+            # TODO : Use logging.debug
+            print("True: Acquiring a new image available\n"
+                  f" Reading buffer number {frame_ind}")
+            er_c, bf_ind, img = self.session.extract_buffer(frame_ind)
+            self.last_measurement[i, :, :] = img
+
+        self.last_frame_acquired = frame_ind
+        self.read_camera_temp()
+
+    def read_camera_temp(self):
+        """
+        Reads the camera temperature, sets self.cameraTemp to the new value
+        """
+        if self.stop_connections or self.reset_connection:
+            return
+
+        if self.enable:
+            msg_in = "?TMP"
+            er_c, msg_out = self.session.hamamatsu_serial(msg_in)
+
+            m = re.match(r"TMP (\d+)\.(\d+)", msg_out)
+            self.camera_temp = float("{}.{}".format(m.group(1), m.group(2)))
+
+        else:
+            self.camera_temp = np.inf
+
+    def data_out(self) -> str:
+        """
+        Returns a formatted string of relevant hamamatsu data to be written to hdf5 fike
+        Returns: formatted data string
+        """
+        if self.stop_connections or self.reset_connection:
+            return ""
+
+        if not self.enable:
+            return ""
+
+        hm = "Hamamatsu"
+        hm_str = ""
+        sz = self.last_measurement.shape
+        hm_str += format_data(f"{hm}/numShots", f"{sz[0]}")
+        hm_str += format_data(f"{hm}/rows", f"{sz[1]}")
+        hm_str += format_data(f"{hm}/columns", f"{sz[2]}")
+
+        for shot in range(sz[0]):
+            flat_ar = np.reshape(self.last_measurement[shot, :, :], sz[1]*sz[2])
+            tmp_str = ar_to_str(flat_ar)  # TODO : implement based on encoding in cspy and labview
+            hm_str += format_data(f"{hm}/shots/{shot}", tmp_str)
+
+        hm_str += format_data(f"{hm}/temperature", "{:.3f}".format(self.camera_temp))
+
+        return hm_str
+
+
+def ar_to_str(ar: np.ndarray) -> str:  # TODO : Implement this, place logically
```

Comment:
@c-poole I believe your assistance would be very useful in implementing this function.